File: 07 July/07-16/pigLatinConverter.js
function pigLatin(str) {

  return str
    .split(" ")
    .map(word => {
      if (/^[aeiou]/.test(word)) {
        return `${word}way`;
      }

      const match = word.match(/^[b-df-hj-np-tv-z]+/i);

      if (match) {
        return match[0][0] === match[0][0].toUpperCase() 
          ? `${word[match[0].length].toUpperCase()}${word.slice(match[0].length + 1)}${match[0].toLowerCase()}ay`
          : `${word.slice(match[0].length)}${match[0]}ay`;
      }
    })
    .join(" ");
}

// --- TEST SUITE ---

const testsText = String.raw`
1. pigLatin("universe") should return "universeway".
2. pigLatin("hello") should return "ellohay".
3. pigLatin("hello universe") should return "ellohay universeway".
4. pigLatin("Hello universe") should return "Ellohay universeway".
5. pigLatin("Pig Latin is fun") should return "Igpay Atinlay isway unfay".
6. pigLatin("The quick brown fox jumped over the lazy dog") should return "Ethay uickqay ownbray oxfay umpedjay overway ethay azylay ogday".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 07 July/07-16/pig_latin_converter.py
def pig_latin(s):
    def convert(word):
        if re.match("[aeiou]", word):
            return f"{word}way"

        match = re.match("^[b-df-hj-np-tv-z]+", word, re.IGNORECASE)
        
        if match:
            return \
                f"{word[match.end()].upper()}{word[match.end() + 1:]}{match[0].lower()}ay" if match[0][0] == match[0][0].upper() \
                else f"{word[match.end():]}{match[0]}ay"

    return " ".join(map(convert, s.split(" ")))

# --- TEST SUITE ---

tests_text = r'''
1. pig_latin("universe") should return "universeway".
2. pig_latin("hello") should return "ellohay".
3. pig_latin("hello universe") should return "ellohay universeway".
4. pig_latin("Hello universe") should return "Ellohay universeway".
5. pig_latin("Pig Latin is fun") should return "Igpay Atinlay isway unfay".
6. pig_latin("The quick brown fox jumped over the lazy dog") should return "Ethay uickqay ownbray oxfay umpedjay overway ethay azylay ogday".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)